compiler/rustc_infer/src/infer/mod.rs

```diff
@@ -1977,3 +1989,119 @@ impl<'tcx> fmt::Debug for RegionObligation<'tcx> {
         )
     }
 }
+
+/// Check if a given constant can be evaluated.
+#[instrument(skip(infcx), level = "debug")]
+pub fn is_const_evaluatable<'cx, 'tcx>(
+    infcx: &InferCtxt<'cx, 'tcx>,
+    uv: ty::Unevaluated<'tcx, ()>,
+    param_env: ty::ParamEnv<'tcx>,
+    span: Span,
+) -> Result<(), NotConstEvaluatable> {
```

Comment:
i think that should stay in `rustc_trait_selection`

Reply:
This can't stay in `rustc_trait_selection` if we intend to eventually use the same check for `FailureKind` as suggested [here](https://rust-lang.zulipchat.com/#narrow/stream/260443-project-const-generics/topic/gce.20.60infcx.2Econst_eval_resolve.60/near/283284045), because that would introduce a cyclic dependency on this crate for `rustc_trait_selection`, and `rustc_trait_selection` for `rustc_infer`. I thought here made the most sense since it uses an `InferCtxt` as well.

Reply:
so `FailureKind` should be moved? But the `is_const_evaluatable` function should stay, because that one should only be used during trait selection.